worker/worker-execution/score-worker-execution-impl/src/main/java/io/cloudslang/worker/execution/services/ExecutionServiceImpl.java

```diff
@@ -475,42 +476,38 @@ private static boolean isDebuggerMode(Map<String, Serializable> systemContext) {
     }
 
     private void dumpBusEvents(Execution execution) throws InterruptedException {
-        ArrayDeque<ScoreEvent> eventsQueue = execution.getSystemContext().getEvents();
-        if (eventsQueue == null) {
-            return;
-        }
-        for (ScoreEvent eventWrapper : eventsQueue) {
-            eventBus.dispatch(eventWrapper);
+        final ArrayDeque<ScoreEvent> eventsQueue = execution.getSystemContext().getEvents();
+        if ((eventsQueue != null) && !eventsQueue.isEmpty()) {
+            for (ScoreEvent eventWrapper : eventsQueue) {
+                eventBus.dispatch(eventWrapper);
+            }
+            eventsQueue.clear();
         }
-        eventsQueue.clear();
     }
 
     protected ExecutionStep loadExecutionStep(Execution execution) {
         RunningExecutionPlan runningExecutionPlan;
-        if (execution != null) {
```

Comment:
Even though this is sometimes an "useless" null check and it might hurt performance(a little bit), I still think it would be better to leave the null check there. Just in case you want to avoid any NPE in the future

Reply:
That is the point it is hurting performance, it is hard to squeeze it :)
It is quite hard to grasp, but it is ok if you check in more detail.
There is no functional change between them since if you check all usages there is no point to execution being null, since it will fail earlier. There is only one case when the exception is replaced from _RuntimeException_ to NPE but that is ok as well since _NullPointerException_ extends _RuntimeException_.

Reply:
I understand the situation, it is not that hard to grasp. I also had to handle those NPEs before. What I am saying is that we should follow the Bloch's recommendations about optimization (read item 67 - Effective Java 3rd Edition).  You recommended this book to me, if I am not mistaken
The thing I worry the most is that, even if I understand the situation and the fact that it might not do any harm as I am worried, we do not have any clear performance measurements. To quote Bloch's book: 

> measure performance before and after each attempted optimization. You may be surprised by what you find. Often, attempted optimizations have no measurable effect on performance; sometimes, they make it worse.

Still, if we have clear results that this optimization is beneficial and does not break anything, merge this.